File: src/measure.mjs
// Microbenchmark helper: warm up, run a thunk N times, return timing samples and
// summary stats. Pure + deterministic (timing aside) so it's unit-testable; used
// by the hot-path harness and embeddable by callers.

// Nearest-rank percentile of a pre-sorted, non-empty array.
function percentile(sorted, p) {
  const idx = Math.min(sorted.length - 1, Math.floor((p / 100) * sorted.length));
  return sorted[idx];
}

/**
 * Summarize timing samples (milliseconds) → { n, min, max, mean, median, p95,
 * opsPerSec }. Empty input yields all-zero stats.
 */
export function summarize(samples) {
  if (samples.length === 0) {
    return { n: 0, min: 0, max: 0, mean: 0, median: 0, p95: 0, opsPerSec: 0 };
  }
  const sorted = [...samples].sort((a, b) => a - b);
  const mean = sorted.reduce((a, b) => a + b, 0) / sorted.length;
  return {
    n: sorted.length,
    min: sorted[0],
    max: sorted[sorted.length - 1],
    mean,
    median: percentile(sorted, 50),
    p95: percentile(sorted, 95),
    opsPerSec: mean > 0 ? 1000 / mean : 0,
  };
}

/**
 * Run `thunk` (sync or async) `iters` times after `warmup` warmup runs, timing
 * each measured run with a high-resolution clock.
 * @param {() => unknown | Promise<unknown>} thunk
 * @param {{ iters?: number, warmup?: number, clock?: () => bigint }} [opts]
 * @returns {Promise<{ samples: number[], stats: object }>}
 */
function measureCfg(opts) {
  const iters = opts.iters ?? 50;
  return {
    iters,
    warmup: opts.warmup ?? Math.min(5, iters),
    clock: opts.clock ?? process.hrtime.bigint,
  };
}

export async function measure(thunk, opts = {}) {
  const cfg = measureCfg(opts);
  for (let i = 0; i < cfg.warmup; i++) await thunk();
  const samples = [];
  for (let i = 0; i < cfg.iters; i++) {
    const t0 = cfg.clock();
    await thunk();
    samples.push(Number(cfg.clock() - t0) / 1e6);
  }
  return { samples, stats: summarize(samples) };
}

/** One-line human report for a labeled stats object. */
export function formatStats(label, stats) {
  const median = stats.median.toFixed(3).padStart(9);
  const ops = Math.round(stats.opsPerSec).toString().padStart(8);
  return `${label.padEnd(26)} ${median} ms  ${ops} ops/s  (mean ${stats.mean.toFixed(3)}, p95 ${stats.p95.toFixed(3)}, n=${stats.n})`;
}

File: rust/crates/turbo-surf-core/src/measure.rs
//! Timing-sample summary (port of `src/measure.mjs` `summarize`): min/max/mean/
//! median/p95/ops-per-sec over millisecond samples. Pure + deterministic.

#[derive(Debug, Clone, PartialEq)]
pub struct Stats {
    pub n: usize,
    pub min: f64,
    pub max: f64,
    pub mean: f64,
    pub median: f64,
    pub p95: f64,
    pub ops_per_sec: f64,
}

// Nearest-rank percentile of a pre-sorted, non-empty slice.
fn percentile(sorted: &[f64], p: f64) -> f64 {
    let idx = ((p / 100.0) * sorted.len() as f64).floor() as usize;
    sorted[idx.min(sorted.len() - 1)]
}

/// Summarize millisecond `samples`. Empty input → all-zero stats.
pub fn summarize(samples: &[f64]) -> Stats {
    if samples.is_empty() {
        return Stats {
            n: 0,
            min: 0.0,
            max: 0.0,
            mean: 0.0,
            median: 0.0,
            p95: 0.0,
            ops_per_sec: 0.0,
        };
    }
    let mut sorted = samples.to_vec();
    sorted.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));
    let mean = sorted.iter().sum::<f64>() / sorted.len() as f64;
    Stats {
        n: sorted.len(),
        min: sorted[0],
        max: sorted[sorted.len() - 1],
        mean,
        median: percentile(&sorted, 50.0),
        p95: percentile(&sorted, 95.0),
        ops_per_sec: if mean > 0.0 { 1000.0 / mean } else { 0.0 },
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn empty_is_zero() {
        assert_eq!(summarize(&[]).n, 0);
        assert_eq!(summarize(&[]).ops_per_sec, 0.0);
    }

    #[test]
    fn summary_stats() {
        let s = summarize(&[10.0, 20.0, 30.0, 40.0, 50.0]);
        assert_eq!(s.n, 5);
        assert_eq!(s.min, 10.0);
        assert_eq!(s.max, 50.0);
        assert_eq!(s.mean, 30.0);
        assert_eq!(s.median, 30.0); // nearest-rank: idx floor(0.5*5)=2 → 30
        assert_eq!(s.p95, 50.0); // idx floor(0.95*5)=4 → 50
        assert!((s.ops_per_sec - 1000.0 / 30.0).abs() < 1e-9);
    }

    #[test]
    fn unsorted_input_is_sorted() {
        let s = summarize(&[50.0, 10.0, 30.0]);
        assert_eq!(s.min, 10.0);
        assert_eq!(s.max, 50.0);
    }
}
